src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlayHints/InlayHintService.cs

```diff
@@ -0,0 +1,126 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.AspNetCore.Razor.LanguageServer.Semantic;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.InlayHints;
+
+[Export(typeof(IInlayHintService))]
+[method: ImportingConstructor]
+internal sealed class InlayHintService(IRazorDocumentMappingService documentMappingService) : IInlayHintService
+{
+    private readonly IRazorDocumentMappingService _documentMappingService = documentMappingService;
+
+    public async Task<InlayHint[]?> GetInlayHintsAsync(IClientConnection clientConnection, VersionedDocumentContext documentContext, Range range, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        // We are given a range by the client, but our mapping only succeeds if the start and end of the range can both be mapped
+        // to C#. Since that doesn't logically match what we want from inlay hints, we instead get the minimum range of mappable
+        // C# to get hints for. We'll filter that later, to remove the sections that can't be mapped back.
+        if (!_documentMappingService.TryMapToGeneratedDocumentRange(csharpDocument, range, out var projectedRange) &&
+            !RazorSemanticTokensInfoService.TryGetMinimalCSharpRange(codeDocument, range, out projectedRange))
+        {
+            // There's no C# in the range.
+            return null;
+        }
+
+        // For now we only support C# inlay hints. Once Web Tools adds support we'll need to request from both servers and combine
+        // the results, much like folding ranges.
+        var delegatedRequest = new DelegatedInlayHintParams(
+            Identifier: documentContext.Identifier,
+            ProjectedRange: projectedRange,
+            ProjectedKind: RazorLanguageKind.CSharp
+        );
+
+        var inlayHints = await clientConnection.SendRequestAsync<DelegatedInlayHintParams, InlayHint[]?>(
+            CustomMessageNames.RazorInlayHintEndpoint,
+            delegatedRequest,
+            cancellationToken).ConfigureAwait(false);
+
+        if (inlayHints is null)
+        {
+            return null;
+        }
+
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
+        using var _1 = ArrayBuilderPool<InlayHint>.GetPooledObject(out var inlayHintsBuilder);
+        foreach (var hint in inlayHints)
+        {
+            if (hint.Position.TryGetAbsoluteIndex(csharpSourceText, null, out var absoluteIndex) &&
+                _documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteIndex, out Position? hostDocumentPosition, out _))
```

Comment:
Can this legitimately fail to map back to host or should we at least assert?

Reply:
It's so legitimate I wouldn't even call it a "failure". It is expected that Roslyn will return all hints that are in the range in the generated code, and we are filtering them down to just the ones that are actually present in the user code.